src/main/java/io/jenkins/plugins/opentelemetry/job/log/OtelLogOutputStream.java

```diff
@@ -0,0 +1,98 @@
+import io.opentelemetry.api.trace.propagation.W3CTraceContextPropagator;
+import io.opentelemetry.context.Context;
+import io.opentelemetry.context.propagation.TextMapGetter;
+import io.opentelemetry.sdk.common.Clock;
+import io.opentelemetry.sdk.logs.LogEmitter;
+import org.apache.commons.lang.StringUtils;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * Process the output stream and send it to OpenTelemetry.
+ * TODO support Pipeline Step Context {@link Context} in addition to supporting run root context
+ * See https://github.com/jenkinsci/pipeline-cloudwatch-logs-plugin/blob/pipeline-cloudwatch-logs-0.2/src/main/java/io/jenkins/plugins/pipeline_cloudwatch_logs/CloudWatchSender.java#L162
+ */
+final class OtelLogOutputStream extends LineTransformationOutputStream {
+    public static boolean ENABLE_LOG_FORMATTING = Boolean.valueOf(System.getProperty("pipeline.log.elastic.enable.log.formatting", "false"));
+    private final static Logger LOGGER = Logger.getLogger(OtelLogOutputStream.class.getName());
+
+    @Nonnull
+    final BuildInfo buildInfo;
+    /**
+     * {@link Map} version of the {@link Context} used to associate log message with the right {@link Span}
+     */
+    final Map<String, String> w3cTraceContext;
+    final LogEmitter logEmitter;
+    final Context context;
+    final Clock clock;
+
+    /**
+     * @param buildInfo
+     * @param w3cTraceContext Serializable version of the {@link Context} used to associate log messages with {@link io.opentelemetry.api.trace.Span}s
+     */
+    public OtelLogOutputStream(BuildInfo buildInfo, Map<String, String> w3cTraceContext, LogEmitter logEmitter, Clock clock) {
+        this.buildInfo = buildInfo;
+        this.logEmitter = logEmitter;
+        this.clock = clock;
+        this.w3cTraceContext = w3cTraceContext;
+        this.context = W3CTraceContextPropagator.getInstance().extract(Context.current(), this.w3cTraceContext, new TextMapGetter<Map<String, String>>() {
+            @Override
+            public Iterable<String> keys(Map<String, String> carrier) {
+                return carrier.keySet();
+            }
+
+            @Nullable
+            @Override
+            public String get(@Nullable Map<String, String> carrier, String key) {
+                return carrier == null ? null : carrier.get(key);
+            }
+        });
+    }
+
+    @Override
+    protected void eol(byte[] bytes, int len) throws IOException {
+        if (len == 0) {
+            return;
+        }
+        ConsoleNotes.TextAndAnnotations textAndAnnotations = ConsoleNotes.parse(bytes, len);
+        String plainLogLine = textAndAnnotations.text;
+        if (plainLogLine == null || plainLogLine.isEmpty()) {
+            LOGGER.log(Level.FINER, () -> buildInfo + " - skip empty log line");
```

Comment:
Could be confusing if you actually had some tool run as part of a `sh` step which deliberately emitted blank lines as a visual separator.